### PR TITLE
Vocabulary Log Probabilities API

### PR DESCRIPTION
**Pull request by Dokotela**
# Vocabulary Log Probabilities API

## Summary

I needed this for fusion from two different models, and its worked ok for me so far. I think there was also another PR/Issues that was asking about this. I tried to add API for accessing full vocabulary log probability, to use with advanced confidence scoring, entropy-based uncertainty estimation, and hypothesis fusion algorithms (e.g., weighted ROVER). This exposes the **complete probability distribution** over the entire vocabulary for each decoded token position.

### Core API (`sherpa-onnx/c-api/`)

**New C API Functions:**
- `SherpaOnnxOnlineStreamGetVocabLogProbs()` - full vocab log probs  online streams
- `SherpaOnnxOfflineStreamGetVocabLogProbs()` - full vocab log probs offline streams  
- `SherpaOnnxDestroyVocabLogProbs()` - Memory management for vocab log probs

**New Data Structure:**
```c
typedef struct SherpaOnnxVocabLogProbs {
  const float *log_probs;  // Flattened 2D array [num_tokens][vocab_size]
  int32_t num_tokens;
  int32_t vocab_size;
} SherpaOnnxVocabLogProbs;
```

### Model Support

- **Online Models:**
  - Streaming Transducer (Zipformer, NeMo)
  - Streaming CTC models

- **Offline Models:**
  - Whisper (greedy search decoder)
  - Moonshine (greedy search decoder)
  - Canary (transducer models)
  - CTC models

### Implementation Details

**Core Changes:**
- Added `vocab_log_probs` field to `OfflineRecognitionResult` and `OnlineRecognizerResult` structures
- Modified decoders to capture full vocabulary distributions during decoding:
  - `offline-whisper-greedy-search-decoder.cc` - Captures joiner output logits
  - `offline-moonshine-greedy-search-decoder.cc` - Captures model output distributions
  - `online-transducer-greedy-search-decoder.cc` - Captures joiner logits per frame
  - `online-transducer-greedy-search-nemo-decoder.cc` - NeMo-specific implementation
  - `offline-recognizer-canary-impl.h` - Canary model support
  - `offline-recognizer-ctc-impl.h` - CTC model support

**Flutter/Dart Bindings:**
- Added `getVocabLogProbs()` methods to `OnlineStream` and `OfflineStream`
- Updated FFI bindings in `sherpa_onnx_bindings.dart`
- Full type-safe access from Dart code

**JSON Serialization:**
- Added `token_log_probs` field to JSON output (for backward compatibility)
- Maintains compatibility with upstream's `ys_log_probs` field

## Summary by CodeRabbit

* **New Features**
  * Added retrieval of vocabulary-level log probabilities for online and offline streams via new getVocabLogProbs().
  * Recognition results now include per-token confidence scores and per-token vocabulary probability distributions.
  * Introduced temperature scaling in the online decoder to control output sharpness.

### CODE REVIEW

**Comment by csukuangfj** [comment]
Please leave a message when it is ready to merge.

File: .github/scripts/test-python.sh [hunk 1]
```diff
@@ -114,7 +114,7 @@ curl -SL -O https://github.com/k2-fsa/sherpa-onnx/releases/download/speaker-segm
 tar xvf sherpa-onnx-pyannote-segmentation-3-0.tar.bz2
 rm sherpa-onnx-pyannote-segmentation-3-0.tar.bz2
 
-curl -SL -O https://github.com/k2-fsa/sherpa-onnx/releases/download/speaker-recongition-models/3dspeaker_speech_eres2net_base_sv_zh-cn_3dspeaker_16k.onnx
```
Comment (csukuangfj) [inline .github/scripts/test-python.sh:230]: This one with the typo is the correct URL. Please see
https://github.com/k2-fsa/sherpa-onnx/releases/tag/speaker-recongition-models

**Comment by Dokotela** [comment]
I think its ready to merge. The failing tests don't seem to have anything to do with my PR (as far as I can tell), and I'm not sure why on this one instance the style_check didn't complete.

**Comment by Dokotela** [comment]
I don't know what a style check is, but otherwise I think this PR is ready.

File: sherpa-onnx/csrc/offline-moonshine-greedy-search-decoder.cc [hunk 2]
```diff
@@ -59,12 +62,36 @@ OfflineMoonshineGreedySearchDecoder::Decode(Ort::Value encoder_out) {
...
+    full_vocab_probs[0] = max_log_prob;
+
+    for (int32_t j = 1; j < vocab_size; ++j) {
+      float log_prob = p[j] - log_sum;
+      full_vocab_probs[j] = log_prob;
+      if (log_prob > max_log_prob) {
+        max_log_prob = log_prob;
+        max_token_id = j;
+      }
+    }
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/offline-moonshine-greedy-search-decoder.cc]: Too complicated. Please use existing functions to simplify it.

File: sherpa-onnx/csrc/offline-recognizer-canary-impl.h [hunk 3]
```diff
@@ -56,6 +57,13 @@ class OfflineRecognizerCanaryImpl : public OfflineRecognizerImpl {
   void DecodeStream(OfflineStream *s) const {
     auto meta = model_->GetModelMetadata();
     auto enc_out = RunEncoder(s);
+
+    if (enc_out.empty()) {
+      OfflineRecognitionResult empty_result;
+      s->SetResult(empty_result);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/offline-recognizer-canary-impl.h]:
```suggestion
```

Just return is enough. Every stream has a default initialized result.

**Review by csukuangfj — CHANGES_REQUESTED** [review]
You have a lot of redundant code. Please try to use existing functions to simplify it.

File: sherpa-onnx/csrc/offline-ctc-greedy-search-decoder.cc [hunk 4]
```diff
@@ -32,16 +32,18 @@ std::vector<OfflineCtcDecoderResult> OfflineCtcGreedySearchDecoder::Decode(
...
-      auto y = static_cast<int64_t>(std::distance(
+      auto max_it = std::max_element(
           static_cast<const float *>(p_log_probs),
-          std::max_element(
-              static_cast<const float *>(p_log_probs),
-              static_cast<const float *>(p_log_probs) + vocab_size)));
+          static_cast<const float *>(p_log_probs) + vocab_size);
+      auto y = static_cast<int64_t>(std::distance(
+          static_cast<const float *>(p_log_probs), max_it));
+      float log_prob = *max_it;
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/offline-ctc-greedy-search-decoder.cc]: Can you use the following function 
https://github.com/k2-fsa/sherpa-onnx/blob/b95ec53293274e85732d544e6e725296ab619a87/sherpa-onnx/csrc/math.h#L170

Once you get the max element index, you can use p_log_probs[index] to get the log prob.

The current code is a bit complex.

File: sherpa-onnx/c-api/c-api.h [hunk 5]
```diff
@@ -1501,6 +1501,21 @@ typedef struct SherpaOnnxOfflineRecognizerResult {
...
+} SherpaOnnxVocabLogProbs;
+
+SHERPA_ONNX_API const SherpaOnnxVocabLogProbs *
+SherpaOnnxOnlineStreamGetVocabLogProbs(const SherpaOnnxOnlineStream *stream);
+
+SHERPA_ONNX_API const SherpaOnnxVocabLogProbs *
+SherpaOnnxOfflineStreamGetVocabLogProbs(const SherpaOnnxOfflineStream *stream);
+
+SHERPA_ONNX_API void SherpaOnnxDestroyVocabLogProbs(
+    const SherpaOnnxVocabLogProbs *log_probs);
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]: Can you put the log_probs inside the result struct, like how we return timestamps?

File: flutter/sherpa_onnx/lib/src/sherpa_onnx_bindings.dart [hunk 6]
```diff
@@ -824,6 +824,61 @@ final class SherpaOnnxSpokenLanguageIdentificationResult extends Struct {
 
 final class SherpaOnnxSpokenLanguageIdentification extends Opaque {}
 
+/// FFI struct for the online recognizer result (mirrors C API struct layout).
```
Comment (csukuangfj) [inline flutter/sherpa_onnx/lib/src/sherpa_onnx_bindings.dart:827]: You don't need to add a new `class`.

Please use our existing method to parse the result, i.e., parse the json result.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

**Comment by Dokotela** [comment]
Closing this.

I built this API to support confidence-weighted ROVER fusion for clinical ASR — the idea being that richer per-token distributions should give better voting than flat equal-weight voting.

After testing seven fusion methods on clinical conversation audio (naive ROVER, CNC with epsilon tuning, confidence-weighted, Shannon/Tsallis entropy-weighted, margin-weighted, and a learned per-word GBM classifier with 5-fold cross-validation), the methods that actually consume vocab log-probs all landed within ~0.1pp of naive equal-weight ROVER (13.30–13.41% vs. 13.41% WER on my datasets). The best practical result came from a method that doesn't use  vocab log-probs at all (epsilon-tuned CNC, 13.05%). The oracle upper bound is 9.07% — a ~4pp gap no voting scheme we tested could close, because roughly two-thirds of the best model's deletions are also errors in every other model in our set. The residual appears to be a structural floor rather than something fusion can fix.

Now I am using this for medical audio, and I suppose other datasets could give different results. However, since I can't demonstrate a measurable benefit for my own motivating use case, I decided to close this.

I'll keep the branch in the very off chance someone else finds something about it useful.

Thanks to the maintainers for the review attention along the way.
